spark_log_parser/parsing_models/application_model_v2.py

```diff
@@ -1,131 +1,368 @@
-        val2.validate()
+        logging.info("Saved object to cloud: %s" % (key))
 
-    def getRecentEvents(self):
 
-        tcomp = np.concatenate(([0.0], self.taskData["end_time"].values))
+DataType = TypeVar("DataType")
+
+
+class AbstractSparkApplicationDataLoader(abc.ABC, Generic[DataType], DataLoader):
+    """
+    Defines the methods that other data loaders should implement in order to appropriately construct
+    some SparkApplication. The order in which these methods are called is defined in construct_spark_application
+    """
+
+    @abc.abstractmethod
+    async def load_raw_datas(self, keys) -> list[DataType]:
+        """
+        Method that should return the data that is our "source-of-truth" (i.e. that data from which we will be
+        constructing the SparkApplication)
+        """
+
+    @abc.abstractmethod
+    def compute_sql_info(self, raw_data: DataType, spark_app: SparkApplication) -> SparkApplication:
+        """ TODO """
+
+    @abc.abstractmethod
+    def compute_executor_info(
+        self, raw_data: DataType, spark_app: SparkApplication
+    ) -> SparkApplication:
+        """ TODO """
+
+    @abc.abstractmethod
+    def compute_all_job_data(
+        self, raw_data: DataType, spark_app: SparkApplication
+    ) -> SparkApplication:
+        """ TODO """
+
+    @abc.abstractmethod
+    def compute_all_task_data(
+        self, raw_data: DataType, spark_app: SparkApplication
+    ) -> SparkApplication:
+        """ TODO """
+
+    @abc.abstractmethod
+    def compute_all_stage_data(
+        self, raw_data: DataType, spark_app: SparkApplication
+    ) -> SparkApplication:
+        """ TODO """
+
+    @abc.abstractmethod
+    def compute_all_driver_accum_data(
+        self, raw_data: DataType, spark_app: SparkApplication
+    ) -> SparkApplication:
+        """ TODO """
+
+    @abc.abstractmethod
+    def compute_all_metadata(
+        self, raw_data: DataType, spark_app: SparkApplication
+    ) -> SparkApplication:
+        """ TODO """
+
+    @abc.abstractmethod
+    def compute_recent_events(
+        self, raw_data: DataType, spark_app: SparkApplication
+    ) -> SparkApplication:
+        """ TODO """
+
+    def construct_spark_application(self, raw_data) -> SparkApplication:
+        """
+        Generic 'recipe' for constructing a SparkApplication from some raw source of data.
+        """
+        spark_app = SparkApplication()
+        spark_app = self.compute_sql_info(raw_data, spark_app)
+        spark_app = self.compute_executor_info(raw_data, spark_app)
+        spark_app = self.compute_all_job_data(raw_data, spark_app)
+        spark_app = self.compute_all_task_data(raw_data, spark_app)
+        spark_app = self.compute_all_stage_data(raw_data, spark_app)
+        spark_app = self.compute_all_driver_accum_data(raw_data, spark_app)
+        spark_app = self.compute_all_metadata(raw_data, spark_app)
+        spark_app = self.compute_recent_events(raw_data, spark_app)
+        return spark_app
```

Comment:
Yeah, I agree that this isn't quite right - I was rethinking this a bit the other day as well. When you say "initialization function", do you mean instead of calling `SparkApplication()` directly? Some `spark_app = initialize_spark_app()`? Or are you thinking of something a bit more comprehensive?

Reply:
`__init__` is probably the most simple alternative...

Reply:
ah - I'm not sure if using `__init__` is viable to construct the initial `SparkApplication()` instance, since this class is intended to be able to instantiate many `SparkApplication`s, if necessary.

Reply:
I meant `SparkApplication.__init__`

Reply:
Ah, gotcha! The reason I haven't done that is because the implementations of these `compute_` methods differ based on whether or not the `raw_data` is a parsed or unparsed log (and in the future, whether or not it is being sourced from e.g. the data platform). The idea was to remove this logic from `SparkApplication` itself, and basically just have it be a "struct", since we don't necessarily want `SparkApplication` to know how to interpret many different sources of `raw_data`. That said, perhaps `SparkApplication` could take the concrete `Loader` instance and just call these methods on it instead 🤔 we just want some way to be able to override this "recipe" if it's not appropriate/necessary for some ways of sourcing data.

Reply:
I can't pass an `AbstractSparkApplicationDataLoader` into `SparkApplication.__init__` without running into circular declaration issues :/ or rather, I _can_, but I can't use type annotations. Basically, given something like this -

```python
class SparkApplication():

  def __init__(self, app_loader: AbstractSparkApplicationDataLoader):
    ...


class AbstractSparkApplicationDataLoader():

  async def batch_load_fn(self, keys) -> list[SparkApplication]:
    ...
```

there's no way we can resolve this circular dependency (at least, not a way that's apparent to me yet!).

Thinking about it more, though, I'm also not sure if we actually _want_ `SparkApplication.__init__` to handle this - the idea behind putting this logic into the `DataLoader`, as opposed to the `SparkApplication`, is that it allows for greater flexibility in how this data gets set on the `SparkApplication`, and `SparkApplication` really just becomes an interface/data container. If `SparkApplication` is opinionated about how this data is set on itself, then we lose that flexibility. We don't want concrete `DataLoader`s to have to "amend"/"correct" anything about a `SparkApplication` if what is happening in `SparkApplication` doesn't actually work (or is unnecessary) for a given scenario.